src/lib/audio/audio-buffer-player.js

```diff
@@ -6,15 +6,40 @@ class AudioBufferPlayer {
         this.buffer = this.audioContext.createBuffer(1, samples.length, this.audioContext.sampleRate);
         this.buffer.getChannelData(0).set(samples);
         this.source = null;
+
+        this.startTime = null;
+        this.updateCallback = null;
+        this.trimStart = null;
+        this.trimEnd = null;
     }
 
-    play (onEnded) {
-        // Buffer source nodes are one time use only. Must do this every play.
+    play (trimStart, trimEnd, onUpdate, onEnded) {
+        this.updateCallback = onUpdate;
+        this.trimStart = trimStart;
+        this.trimEnd = trimEnd;
+        this.startTime = Date.now();
+
+        const trimStartTime = this.buffer.duration * trimStart;
+        const trimmedDuration = this.buffer.duration * trimEnd - trimStartTime;
+
         this.source = this.audioContext.createBufferSource();
         this.source.onended = onEnded;
         this.source.buffer = this.buffer;
         this.source.connect(this.audioContext.destination);
-        this.source.start();
+        this.source.start(0, trimStartTime, trimmedDuration);
+
+        this.update();
+    }
+
+    update () {
+        const timeSinceStart = (Date.now() - this.startTime) / 1000;
+        const percentage = timeSinceStart / this.buffer.duration;
+        if (percentage + this.trimStart < this.trimEnd && this.source.onended) {
+            requestAnimationFrame(this.update.bind(this));
+            this.updateCallback(percentage + this.trimStart);
+        } else {
+            this.updateCallback = null;
+        }
     }
 
     stop () {
```

Comment:
Since those are just values, I was thinking they are ok to just leave to get normally garbage collected. The only reason I did some manual cleanup for certain things is because if you leave callback references to react components around, they can't be garbage collected and can cause you problems.